### PR TITLE
[WIP] Add experimental QUIC support

### PR DESCRIPTION
**Pull request by lucas-clemente**
This PR adds experimental QUIC support, as discussed in #47.

### CODE REVIEW

**Comment by lucas-clemente** [comment]
You can test this PR by starting caddy with the `-quic` flag, and then running Chrome with something like this:

```
/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --user-data-dir=/tmp/chrome --no-proxy-server --enable-quic --origin-to-force-quic-on=quic.clemente.io:443 --host-resolver-rules='MAP quic.clemente.io:443 127.0.0.1:2020' 'https://quic.clemente.io/'
```

for the caddyfile

```
https://quic.clemente.io:2020
tls /etc/letsencrypt/live/quic.clemente.io/cert.pem /etc/letsencrypt/live/quic.clemente.io/privkey.pem
```

**Comment by mholt** [comment]
Awesome!

I'm getting `ERR_QUIC_PROTOCOL_ERROR` -- I followed your instructions exactly but changed `quic.clemente.io` to a domain I already have a certificate for.

Also, try building Caddy with the `-race` flag and you'll see a couple race conditions as you start the server and then again when you try to make a request. (That's probably a separate, known issue, but maybe it has something to do with the fact that it works for you but not for me?)

**Comment by lucas-clemente** [comment]
Hm, interesting. Sorry for the troubles! Several things come to mind:

- Could you please double-check that you changed the domain everywhere? It 3 times in the Chrome command, and again in the caddyfile.
- Could you add `utils.SetLogLevel(utils.LogLevelDebug)` and import `github.com/lucas-clemente/quic-go/utils` anywhere in caddy, then post the log here? (This is not intended to be a public API, but may help some.)
- Third, when you go to chrome://net-internals/#events you should see some QUIC events marked in red, could you post their logs here?

I see the race conditions and will investigate them in https://github.com/lucas-clemente/quic-go/issues/157.

**Comment by mholt** [comment]
Got it! Thanks for the logging tip.

Ahhh, it seems that I am using an ECC certificate which is not supported yet:

```
Closing session with error: only RSA keys are supported for now
Internal error: only RSA keys are supported for now
-> Sending packet 0x1 (65 bytes)
	-> &frames.ConnectionCloseFrame{ErrorCode:0x1, ReasonPhrase:"only RSA keys are supported for now"}
```

Trying again with an RSA certificate, I see:

```
	<- &frames.AckFrame{LargestObserved:0x4, Entropy:0x6, NackRanges:[]frames.NackRange(nil), Truncated:false, DelayTime:579584000, PacketReceivedTime:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}
	<- &frames.StopWaitingFrame{LeastUnacked:0x3, Entropy:0x2, PacketNumberLen:0x0, PacketNumber:0x0}
	<- &frames.ConnectionCloseFrame{ErrorCode:0x2a, ReasonPhrase:"Proof invalid: Failed to verify certificate chain: net::ERR_CERT_AUTHORITY_INVALID"}
Closing session with error: ProofInvalid: Proof invalid: Failed to verify certificate chain: net::ERR_CERT_AUTHORITY_INVALID
```

I admit my certificates have been from Let's Encrypt's staging server while I've been in dev, so they aren't trusted. (I don't know much about QUIC but I'm surprised that causes the connection to fail entirely.)

So I changed to using the real LE endpoint and got a real certificate. Success!

screen shot 2016-05-31 at 10 16 20 am

Thanks again for your work. Looking forward to making this become mainstream!

File: caddyhttp/httpserver/server.go [hunk 1]
```diff
@@ -187,7 +196,20 @@ func (s *Server) Serve(ln net.Listener) error {
 		go runTLSTicketKeyRotation(s.Server.TLSConfig, timer, s.tlsGovChan)
 	}
 
-	return s.Server.Serve(ln)
+	if QUIC {
+		go func() {
+			err := s.quicServer.ListenAndServe()
+			if err != nil {
+				fmt.Printf("Error listening for QUIC connections: %s", err.Error())
```
Comment (mholt) [inline caddyhttp/httpserver/server.go]: FYI I changed this to `log.Printf` in my version.

**Comment by lucas-clemente** [comment]
Very cool 👍

Yeah, I think it's weird that Chrome fails with an internal error when it can't verify the certificate. I think this is a bug.

I've added an issue for ECC private keys: https://github.com/lucas-clemente/quic-go/issues/158

**Comment by lucas-clemente** [comment]
One more thing: I've just added a function [`SetQuicHeaders`](https://godoc.org/github.com/lucas-clemente/quic-go/h2quic#Server.SetQuicHeaders) to the `h2quic.Server`, which sets the appropriate headers so that browser know the server talks QUIC. I've pushed an updated commit where the function is properly called :)

**Comment by elcore** [comment]
Well done @lucas-clemente -- I :heart: your work

> **Be the change that you wish to see in the world.**

Mahatma Gandhi

**Comment by abiosoft** [comment]
Thanks Lucas 😃

**Comment by lucas-clemente** [comment]
Also thank @marten-seemann, he did the other half of the work :)

**Comment by lucas-clemente** [comment]
@mholt: ECC should be supported now, if you get a chance, retry and let me know whether it worked.

**Comment by mholt** [comment]
@lucas-clemente ECC works like a charm :+1:

**Comment by mholt** [comment]
@lucas-clemente I'm going to close this now since I'll be opening a new branch for 0.9 very soon, and it includes your changes here. Thanks for the great work!!

**Comment by mholt** [comment]
@lucas-clemente Just FYI, the 0.9-wip branch is now up, and instructions are added to the Caddy wiki: https://github.com/mholt/caddy/wiki/QUIC

**Pull request closed without merging**